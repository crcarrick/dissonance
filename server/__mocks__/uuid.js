export const v1 = jest.fn();
export const v4 = jest.fn();
