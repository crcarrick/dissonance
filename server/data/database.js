import { Model, Sequelize } from 'sequelize';

import { channel } from './channel/channel.model';
import { message } from './message/message.model';
import { server } from './server/server.model';
import { user } from './user/user.model';

export const connectDatabase = () => {
  const sequelize = new Sequelize(
    process.env.DATABASE_NAME,
    process.env.DATABASE_USERNAME,
    process.env.DATABASE_PASSWORD,
    {
      host: 'localhost',
      dialect: 'postgres',
      logging: false,
    }
  );

  const Channel = channel({ sequelize });
  const Message = message({ sequelize });
  const Server = server({ sequelize });
  const User = user({ sequelize });

  class UserServer extends Model {}

  UserServer.init({}, { sequelize, tableName: 'users_servers' });

  const models = { Channel, Message, Server, User, UserServer };

  Object.keys(models).forEach((key) => {
    if ('associate' in models[key]) {
      models[key].associate(models);
    }
  });

  return { models, sequelize };
};
