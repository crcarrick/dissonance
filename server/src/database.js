import { merge } from 'lodash';

import knexfile from '../knexfile';

export const createDbClient = (knex, database) =>
  database
    ? knex(merge(knexfile[process.env.NODE_ENV], { connection: { database } }))
    : knex(knexfile[process.env.NODE_ENV]);
