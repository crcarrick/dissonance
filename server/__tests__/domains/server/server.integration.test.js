import { gql } from 'apollo-server';
import aws from 'aws-sdk';
import { v4 as uuidv4 } from 'uuid';

import { createDbClient } from '@dissonance/database';

import {
  createTestClient,
  generateMockData,
  seedDatabase,
} from '@dissonance/test-utils';

const knex = jest.requireActual('knex');

describe('Server Integration', () => {
  const { users, servers } = generateMockData();

  let dbClient;
  beforeAll(() => {
    dbClient = createDbClient(
      knex,
      `dissonance_test_${process.env.JEST_WORKER_ID}`
    );
  });

  afterAll(async () => {
    await dbClient.destroy();
  });

  let gqlClient;
  beforeEach(async () => {
    await seedDatabase(dbClient);

    gqlClient = createTestClient({
      context: {
        user: users[0],
      },
      dbClient,
    });
  });

  test('gets a server', async () => {
    const response = await gqlClient.query({
      query: gql`
        query GetServer($input: GetServerInput!) {
          server(input: $input) {
            id
            name
          }
        }
      `,
      variables: {
        input: {
          id: servers[0].id,
        },
      },
    });

    expect(response.data.server).toMatchSnapshot();
  });

  test('gets all servers', async () => {
    const response = await gqlClient.query({
      query: gql`
        query GetServers {
          servers {
            id
            name
          }
        }
      `,
    });

    expect(response.data.servers).toMatchSnapshot();
  });

  test('creates a signed url for avatar upload when provided an image', async () => {
    aws.S3Mock.getSignedUrlPromise.mockReturnValueOnce('https://www.test.com');
    uuidv4.mockReturnValueOnce('1234');

    const response = await gqlClient.mutate({
      mutation: gql`
        mutation CreateServerAvatarSignedUrl(
          $input: CreateServerSignedUrlInput!
        ) {
          createServerAvatarSignedUrl(input: $input) {
            signedUrl
            url
          }
        }
      `,
      variables: {
        input: {
          fileName: 'test.png',
          serverId: servers[0].id,
        },
      },
    });

    expect(response.data.createServerAvatarSignedUrl).toMatchSnapshot();
  });

  test('does not create a signed url for avatar upload when provided a non-image', async () => {
    const response = await gqlClient.mutate({
      mutation: gql`
        mutation CreateServerAvatarSignedUrl(
          $input: CreateServerSignedUrlInput!
        ) {
          createServerAvatarSignedUrl(input: $input) {
            signedUrl
            url
          }
        }
      `,
      variables: {
        input: {
          fileName: 'test.pdf',
          serverId: servers[0].id,
        },
      },
    });

    expect(response.errors).toMatchSnapshot();
  });

  test('creates a server', async () => {
    const response = await gqlClient.mutate({
      mutation: gql`
        mutation CreateServer($input: CreateServerInput!) {
          createServer(input: $input) {
            id
            name
          }
        }
      `,
      variables: {
        input: {
          name: 'Testing 123',
        },
      },
    });

    expect(response.data.createServer).toMatchSnapshot({
      id: expect.any(String),
    });
  });

  test('deletes a server if authorized', async () => {
    const response = await gqlClient.mutate({
      mutation: gql`
        mutation DeleteServer($input: DeleteServerInput!) {
          deleteServer(input: $input) {
            id
          }
        }
      `,
      variables: {
        input: {
          id: servers[0].id,
        },
      },
    });

    expect(response.data.deleteServer).toMatchSnapshot();
  });

  test('does not delete a server if unauthorized', async () => {
    const response = await gqlClient.mutate({
      mutation: gql`
        mutation DeleteServer($input: DeleteServerInput!) {
          deleteServer(input: $input) {
            id
          }
        }
      `,
      variables: {
        input: {
          id: servers[1].id,
        },
      },
    });

    expect(response.errors).toMatchSnapshot();
  });
});
