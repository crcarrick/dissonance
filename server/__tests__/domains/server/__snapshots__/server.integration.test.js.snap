// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server Integration creates a server 1`] = `
Object {
  "id": Any<String>,
  "name": "Testing 123",
}
`;

exports[`Server Integration creates a signed url for avatar upload when provided an image 1`] = `
Object {
  "signedUrl": "https://www.test.com/",
  "url": "https://dissonance.s3.amazonaws.com/1234",
}
`;

exports[`Server Integration deletes a server if authorized 1`] = `
Object {
  "id": "029beee1-5599-4c4b-880d-c238b3ab3d16",
}
`;

exports[`Server Integration does not create a signed url for avatar upload when provided a non-image 1`] = `
Array [
  [GraphQLError: File is not an image type],
]
`;

exports[`Server Integration does not delete a server if unauthorized 1`] = `
Array [
  [GraphQLError],
]
`;

exports[`Server Integration gets a server 1`] = `
Object {
  "id": "029beee1-5599-4c4b-880d-c238b3ab3d16",
  "name": "Inventore provident est",
}
`;

exports[`Server Integration gets all servers 1`] = `
Array [
  Object {
    "id": "029beee1-5599-4c4b-880d-c238b3ab3d16",
    "name": "Inventore provident est",
  },
  Object {
    "id": "1267b1ac-3833-420f-b8a4-a4f25dbbc73f",
    "name": "Impedit facere harum",
  },
  Object {
    "id": "df5bb0e8-1143-459f-8fa2-cf5034a55764",
    "name": "Non quos possimus",
  },
]
`;
